src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticTaggerWrapper.cs

```diff
@@ -100,7 +103,14 @@ public void Dispose()
         public async Task WaitForTags()
         {
             await _listenerProvider.GetWaiter(FeatureAttribute.Workspace).ExpeditedWaitAsync();
-            await _listenerProvider.GetWaiter(FeatureAttribute.SolutionCrawler).ExpeditedWaitAsync();
+
+            // Wait for incremental analyzer until completion.
+            // Ideally, we would we able to "await _listenerProvider.GetWaiter(FeatureAttribute.SolutionCrawler).ExpeditedWaitAsync();"
+            // but that seems to not complete analyzers in presence of cancellation.
+            // See https://github.com/dotnet/roslyn/issues/46786 for intermittent test failures in presence of cancellation.
+            if (!_incrementalAnalyzers.IsEmpty)
+                _registrationService.GetTestAccessor().WaitUntilCompletion(_workspace, _incrementalAnalyzers);
```

Comment:
This change takes the tests out of the production code paths. It would be better to temporarily disable the tests (if necessary) than to "fix" them by running them against non-production code.

Reply:
💡 If you could instead tie this to the production code mechanism that Run Code Analysis on Solution uses, it would be better.

Reply:
Where does Diagnostic tagger explicitly fetch the listener for SolutionCrawler and wait for me? I believe the test code is already not completely in sync with production. Additionally, relying on production code for validating functional aspect seems reasonable - which we already do with Tomas change to now use the IDE's diagnostic service and solution crawler mechanism to rely on diagnostics. This change basically reduces the noise from cancellation, which is switched to CancellationToken.None for majority of our tests, which doesn't mean they are not testing the production code path.